### ISSUE TITLE
[Feature Request] Common output manager for all threads

### ISSUE THREAD

**Issue by redbaron4** [snort3/snort3#197]
Snort++ is multi-threaded and each thread generates its own output (in a separate file/directory depending on configuration). This creates a difficult situation when using the dumped information as Barnyard2 can't handle multiple output unified files unless each file is stored under a different sensor in DB.

Zeek solves this problem by having all logs come to a separate process that only collects logs/outputs from diverse workers and writes a consolidated output file for each event. Would Snort consider moving towards such an architecture so that number of threads is transparent to end user as far as outputs go?

**Comment by snortadmin** [comment]
We are considering logger updates wrt threading and separating format from output, but the general problem of knowning to which file an event was logged is likely to remain outside Snort, at least for the foreseeable future. Have you looked into logstash and similar?

**Comment by redbaron4** [comment]
Thanks for looking at this @snortadmin 

Yes we have considered using Logstash. In fact we can reuse Banryard2 as well but its just that each thread will show up as a separate sensor on the frontend.

Problem with this approach is that the tool configuration needs to be changed each time the number of threads is updated because number/location of "watched" output files changes. By keeping a separate thread for logging, Snort++ can make this transparent to the end user so the user only tracks one unified2/JSON/CSV file irrespective of the threads in use.

**Comment by snortadmin** [comment]
Right. Something we've discussed but not a priority at the moment. We'd likely need a pool of logger threads since one won't always be sufficient. In any case, the number of packet threads shouldn't change once a deployment is figured out.

FYI: unified2 / barnyard is legacy logging and not actively developed. Barnyard2 doesn't support the Snort 2 extra data that was added years ago. And we aren't planning updates to the Snort 3 unified2 logger.

**Comment by redbaron4** [comment]
Thanks for the info re the issue.

As for unified2 is there a recommended logger now that unified2 is deprecated? We are considering `alert_json` if that is stable enough.

**Comment by snortadmin** [comment]
alert_json is a good choice. You can get packet data in base64. The format has more overhead than u2 but is much more flexible and there are tools available for aggregation and transformation, etc. We are making all new fields of interest available via json and csv.
